lms/services/course_roster.py

```diff
@@ -1,49 +1,140 @@
+
         lti_registration = self._db.scalars(
             select(LTIRegistration)
             .join(ApplicationInstance)
-            .join(LMSCourseApplicationInstance)
             .where(LMSCourseApplicationInstance.lms_course_id == lms_course.id)
+            .join(LMSCourseApplicationInstance)
             .order_by(LTIRegistration.updated.desc())
         ).first()
-
         assert lti_registration, "No LTI registration found for LMSCourse."
-        assert (
-            lms_course.lti_context_memberships_url
-        ), "Trying fetch roster for course without service URL."
 
         roster = self._lti_names_roles_service.get_context_memberships(
             lti_registration, lms_course.lti_context_memberships_url
         )
-        LOG.info(
-            "Roster for %s. Users returned %s",
-            lms_course.h_authority_provided_id,
-            len(roster),
+
+        # Insert any users we might be missing in the DB
+        lms_users_by_lti_user_id = {
+            u.lti_user_id: u
+            for u in self._get_roster_users(
+                roster, lms_course.tool_consumer_instance_guid
+            )
+        }
+        # Also insert any roles we might be missing
+        lti_roles_by_value: dict[str, LTIRole] = {
+            r.value: r for r in self._get_roster_roles(roster)
+        }
+
+        # Make sure any new rows have IDs
+        self._db.flush()
+
+        roster_upsert_elements = []
+
+        for member in roster:
+            lti_user_id = member.get("lti11_legacy_user_id") or member["user_id"]
+            # Now, for every user + role, insert a row  in the roster table
+            for role in member["roles"]:
+                roster_upsert_elements.append(
+                    {
+                        "lms_course_id": lms_course.id,
+                        "lms_user_id": lms_users_by_lti_user_id[lti_user_id].id,
+                        "lti_role_id": lti_roles_by_value[role].id,
+                        "active": member["status"] == "Active",
+                    }
+                )
+        # We'll first mark everyone as non-Active. We keep a record of how belonged to a course.
+        self._db.execute(
+            update(CourseRoster)
+            .where(CourseRoster.lms_course_id == lms_course.id)
+            .values(active=False)
+        )
+
+        # Insert and update roster rows.
+        bulk_upsert(
+            self._db,
+            CourseRoster,
+            values=roster_upsert_elements,
+            index_elements=["lms_course_id", "lms_user_id", "lti_role_id"],
+            update_columns=["active", "updated"],
         )
 
+    def _get_roster_users(self, roster, tool_consumer_instance_guid):
+        values = []
+        for member in roster:
+            lti_user_id = member.get("lti11_legacy_user_id") or member["user_id"]
+            name = display_name(
+                member.get("name", ""), member.get("family_name", ""), "", ""
```

Comment:
Perhaps this would benefit from using named parameters, to know what those empty strings are.

Reply:
Yes! Done